### PR TITLE
Clean up flushSync flow types

### PR DESCRIPTION
**Pull request by rickhanlonii**
## Overview

Makes the callback function optional and cleans up the type to make it consistent and correct everywhere.

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberWorkLoop.new.js [hunk 1]
```diff
@@ -1093,10 +1093,7 @@ export function discreteUpdates<A, B, C, D, R>(
   }
 }
 
-export function flushSyncWithoutWarningIfAlreadyRendering<A, R>(
-  fn: A => R,
-  a: A,
```
Comment (rickhanlonii) [inline packages/react-reconciler/src/ReactFiberWorkLoop.new.js:1098]: I think this second param is outdated. It's not used internally anywhere.

File: packages/react-noop-renderer/src/createReactNoop.js [hunk 2]
```diff
@@ -913,7 +913,7 @@ function createReactNoop(reconciler: Function, useMutation: boolean) {
       }
     },
 
-    flushSync(fn: () => mixed) {
+    flushSync<T>(fn: ?() => T): ?T {
```
Comment (acdlite) [inline packages/react-noop-renderer/src/createReactNoop.js:916]: I don't know how to do function overloading in Flow, but it needs to be that if a function is passed, the return type is the same as the return type of the function. As-is, the caller would need to check if the result is undefined.

Comment (rickhanlonii) [reply packages/react-noop-renderer/src/createReactNoop.js]: Yeah good catch. I pushed a way to do it that works.

## Before
Screen Shot 2021-07-15 at 3 46 19 PM

## After
Screen Shot 2021-07-15 at 4 03 51 PM

Comment (rickhanlonii) [reply packages/react-noop-renderer/src/createReactNoop.js]: Note, I tried using an intersection type but that results in some really cryptic errors like "R does not match return type R" so I opted for this version instead.

**Review by acdlite — APPROVED** [review]
Looks good!

**Pull request merged**